packages/xstate-test/src/TestModel.ts

```diff
@@ -0,0 +1,491 @@
+  TestStepResult
+} from './types';
+import { flatten, formatPathTestResult, simpleStringify } from './utils';
+
+export interface TestModelDefaults<TState, TEvent extends EventObject> {
+  coverage: Array<CoverageFunction<TState, TEvent>>;
+  planGenerator: PlanGenerator<TState, TEvent>;
+}
+
+export const testModelDefaults: TestModelDefaults<any, any> = {
+  coverage: [coversAllStates<any, any>(), coversAllTransitions<any, any>()],
+  planGenerator: traverseShortestPlans
+};
+
+/**
+ * Creates a test model that represents an abstract model of a
+ * system under test (SUT).
+ *
+ * The test model is used to generate test plans, which are used to
+ * verify that states in the `machine` are reachable in the SUT.
+ *
+ */
+export class TestModel<TState, TEvent extends EventObject> {
+  private _coverage: TestModelCoverage<TState, TEvent> = {
+    states: {},
+    transitions: {}
+  };
+  public options: TestModelOptions<TState, TEvent>;
+  public defaultTraversalOptions?: TraversalOptions<TState, TEvent>;
+  public getDefaultOptions(): TestModelOptions<TState, TEvent> {
+    return {
+      serializeState: (state) => simpleStringify(state) as SerializedState,
+      serializeEvent: (event) => simpleStringify(event) as SerializedEvent,
+      getEvents: () => [],
+      states: {},
+      events: {},
+      stateMatcher: (_, stateKey) => stateKey === '*',
+      getStates: () => [],
+      execute: () => void 0,
+      logger: {
+        log: console.log.bind(console),
+        error: console.error.bind(console)
+      }
+    };
+  }
+  public static defaults: TestModelDefaults<any, any> = testModelDefaults;
+
+  constructor(
+    public behavior: SimpleBehavior<TState, TEvent>,
+    options?: Partial<TestModelOptions<TState, TEvent>>
+  ) {
+    this.options = {
+      ...this.getDefaultOptions(),
+      ...options
+    };
+  }
+
+  public getPlans(
+    options?: GetPlansOptions<TState, TEvent>
+  ): Array<StatePlan<TState, TEvent>> {
+    const planGenerator = planGeneratorWithDedup<TState, TEvent>(
+      options?.planGenerator || TestModel.defaults.planGenerator
+    );
+    const plans = planGenerator(this.behavior, this.resolveOptions(options));
+
+    return plans;
+  }
+
+  public getShortestPlans(
+    options?: Partial<TraversalOptions<TState, TEvent>>
+  ): Array<StatePlan<TState, TEvent>> {
+    return this.getPlans({ ...options, planGenerator: traverseShortestPlans });
+  }
+
+  public getShortestPlansTo(
+    stateValue: StatePredicate<TState>
+  ): Array<StatePlan<TState, TEvent>> {
+    let minWeight = Infinity;
+    let shortestPlans: Array<StatePlan<TState, TEvent>> = [];
+
+    const plans = this.filterPathsTo(stateValue, this.getShortestPlans());
+
+    for (const plan of plans) {
+      const currWeight = plan.paths[0].weight;
+      if (currWeight < minWeight) {
+        minWeight = currWeight;
+        shortestPlans = [plan];
+      } else if (currWeight === minWeight) {
+        shortestPlans.push(plan);
+      }
+    }
```

Comment:
I don't quite understand this algorithm - I mean I understand the filtering algorithm based on weights but I don't understand the concept of filtering *plans* based on a *single* path of that plan. How do we even guarantee within `@xstate/graph` that the first path is going to have the lowest weight within that plan? 🤔 

Are we first filtering paths within a plan based on weights in `getShortestPlans` and then filter plans here based on those results based on weights again? Why this second filter is not implemented in the `traverseShortestPathsTo`?

Reply:
cc @davidkpiano - I wonder if this still makes sense given that I changed this to paths only now.

Reply:
We can change this to assume an array of paths and find the path satisfying the predicate with the least weight